keps/2724-topology-aware-schedling/README.md

```diff
@@ -463,28 +471,72 @@ at each topology level to the specific subset of nodes.
 type PodSetAssignment struct {
   ...
 
-  // TopologyAssignment indicates the resources assigned per topology level
+  // topologyAssignment indicates the topology assignment divided into
+  // topology domains corresponding to the lowest level of the topology.
+  // The assignment specifies the number of Pods to be scheduled per topology
+  // domain and specifies the node selectors for each topology domain, in the
+  // following way: the node selector keys are specified by the levels field
+  // (same for all domains), and the corresponding node selector value is
+  // specified by the domains.values subfield.
+  //
+  // Example:
+  //
+  // topologyAssignment:
+  //   levels:
+  //   - cloud.provider.com/topology-block
+  //   - cloud.provider.com/topology-rack
+  //   domains:
+  //   - values: [block-1, rack-1]
+  //     count: 4
+  //   - values: [block-1, rack-2]
+  //     count: 2
+  //
+  // Here:
+  // - 4 Pods are to be scheduled on nodes matching the node selector:
+  //   cloud.provider.com/topology-block: block-1
+  //   cloud.provider.com/topology-rack: rack-1
+  // - 2 Pods are to be scheduled on nodes matching the node selector:
+  //   cloud.provider.com/topology-block: block-1
+  //   cloud.provider.com/topology-rack: rack-2
+  //
   // +optional
   TopologyAssignment *TopologyAssignment `json:"topologyAssignment,omitempty"`
 }
 
 type TopologyAssignment struct {
-  // Groups contains the list of assignments split into groups corresponding
-  // to the same topology domain at the lowest level of the hierarchy.
+  // levels is an ordered list of keys denoting the levels of the assigned
+  // topology (i.e. node label keys), from the highest to the lowest level of
+  // the topology.
+  //
   // +required
   // +listType=atomic
   // +kubebuilder:validation:MinItems=1
-  Groups []TopologyAssignmentGroup `json:"groups"`
+  // +kubebuilder:validation:MaxItems=8
+  Levels []string `json:"levels"`
+
+  // domains is a list of topology assignments split by topology domains at
+  // the lowest level of the topology.
+  //
+  // +required
+  Domains []TopologyDomainAssignment `json:"domains"`
 }
 
-type TopologyAssignmentGroup struct {
-  // NodeLabels constitutes the nodeSelector for a given slice of pods. It
-  // defines values for all labels configured in the Topology.Levels.
+type TopologyDomainAssignment struct {
+  // values is an ordered list of node selector values describing a topology
+  // domain. The values correspond to the consecutive topology levels, from
+  // the highest to the lowest.
+  //
+  // +required
+  // +listType=atomic
   // +kubebuilder:validation:MinItems=1
-  NodeLabels map[string]string `json:"nodeLabels"`
+  // +kubebuilder:validation:MaxItems=8
+  Values []string `json:"values"`
```

Comment:
I think the `values` are a slightly generic and It's hard to recognize the objective for this field since the `topologyDomainAssignment` does not have the context for the nodeSelector.

Will we receive the requests except nodeSelector keys here?

Reply:
> > Will we receive the requests except nodeSelector keys here?
> 
> The node selectors are create by combining only the keys (.levels field) and values (.domains.values) fields.

Sorry, I meant the nodeSelector values.

> I agree, this is why I mention the node selector in the comment. We could name the field nodeSelectorValues but it seems verbose.

In that case, How about the `levelValues` since these values are for the levels?

Reply:
> In that case, How about the levelValues since these values are for the levels?

As a name it sgtm, but maybe the only potential downside is grpc size if we have many domains assigned? I guess for big workloads it could be 10k domains or more (if the lowest level is individual node). WDYT?

Reply:
I guess that `levelValues` does not significantly impact on the gRPC message size since the field name increase just 0.05 Mbit in the 10k domains situations.

Additionally, I guess that the TAS will be used for the Top of Rack Switch (ToR) or Leaf or Spine switches instead of Nodes since the topology often needs to be considered based on the network topology, right?

Let me know if you have the actual or imaginable situations where we need to construct the topology based on the Nodes.

Reply:
> I guess that levelValues does not significantly impact on the gRPC message size since the field name increase just 0.05 Mbit in the 10k domains situations.

Yes, it is not much, but the overall API object size in etcd is 1.5Mi which is not much too, so the gain around 3% of the max size.

Reply:
> For example, if you assign at the level of rack, it could happen that a Pod can fit in a rack, but cannot fit on an individual node. I guess it will need to depend on the user preference.

I feel that this usage beyonds the initial TAS motivations. But, I know that we received similar requests here: https://github.com/kubernetes-sigs/kueue/issues/3211

> Yes, it is not much, but the overall API object size in etcd is 1.5Mi which is not much too, so the gain around 3% of the max size.

Yes, that's true. Here, key problem is which disadvantages should we take. I think that both selection have the below disadvantages:

- `values`: This lacks context and background that this field is for the values for nodeSelectors defined in the level. After we read the API document, we can understand the purpose for this field.
- `levelValues`: This will increase the Workload object size. This may cause the etcd performance issues.

Based on the fact that Workload object is for Kueue developer and not exposed to the batch users, I am leaning toward accepting `values`. In that case, could you mention this `values` vs `levelValues` discussion on the Alternative? I guess that we can revisit this alternative based on user feedback like batch admins often struggle caused by this field name?